File: java/exchangeapi/src/test/java/com/thesoftwaregorilla/tdd/money/ExchangeRateAPITest.java
package com.thesoftwaregorilla.tdd.money;

import com.thesoftwaregorilla.helper.DateHelper;
import com.thesoftwaregorilla.helper.JsonHelper;
import com.thesoftwaregorilla.helper.StandardDateFormat;
import com.thesoftwaregorilla.tdd.money.http.GetRequest;
import com.thesoftwaregorilla.tdd.money.jsonholder.ExchangeRateApiPairResponse;
import com.thesoftwaregorilla.tdd.money.jsonholder.ExchangeRateApiStandardResponse;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.text.ParseException;
import java.util.Date;

import static com.thesoftwaregorilla.tdd.money.AppConfigTest.API_PAIR_URL;
import static com.thesoftwaregorilla.tdd.money.AppConfigTest.API_STANDARD_URL;
import static org.junit.jupiter.api.Assertions.*;

//<editor-fold desc="TO DO's">
//TODO: Refactor tests to segregate each class into its own test class
//TODO: Add an CurrencyExchange object to cache all exchange rates and lazy load them as needed.
//TODO: Get to 100% code coverage.
//TODO: Create the utility class for the JSON parsing.
//TODO: Create a utility class for the HTTP requests.
//TODO: Build solution for error responses.
//</editor-fold>

public class ExchangeRateAPITest {

    private static final String BASE_CURRENCY = "USD";
    private static final String TARGET_CURRENCY = "ZAR";
    private static final String ERROR_NO_KEY =
            """
            {"result":"error","error-type":"invalid-key","message":"Invalid API key"}""";
    private static final String ERROR_INVALID_KEY =
            """
            {"result":"error","documentation":"https://www.exchangerate-api.com/docs","terms-of-use":"https://www.exchangerate-api.com/terms","error-type":"invalid-key"}""";
    private static final String STANDARD_RESPONSE_FILE = "src/test/resources/exchange_api_standard_response.json";
    private static final String PAIR_RESPONSE_FILE = "src/test/resources/exchange_api_pair_response.json";

    private static final String API_KEY_PROPERTY;
    private static final String API_PAIR_URL_PROPERTY;
    private static final String API_STANDARD_URL_PROPERTY;
    private static final AppConfig APP_CONFIG;
    static {
        APP_CONFIG = new AppConfig("src/main/resources/application.settings");
        API_KEY_PROPERTY = APP_CONFIG.getApiKey();
        API_PAIR_URL_PROPERTY = APP_CONFIG.getExchangeApiPairUrl();
        API_STANDARD_URL_PROPERTY = APP_CONFIG.getExchangeApiStandardUrl();
    }

    @Test
    public void testHttpGetStandard() {
        GetRequest getRequest = new GetRequest(String.format(API_STANDARD_URL_PROPERTY, API_KEY_PROPERTY, BASE_CURRENCY));
        assertNotNull(getRequest);
        assertEquals(String.format(API_STANDARD_URL, API_KEY_PROPERTY, BASE_CURRENCY), getRequest.getUrl());
        String response = getRequest.fetchResponse();
        assertNotNull(response);
        assertNotEquals(ERROR_INVALID_KEY, response);
        assertNotEquals(ERROR_NO_KEY, response);
//        try{
//            FileHelper.writeStringToFile(response, STANDARD_RESPONSE_FILE);
//        } catch (IOException e) {
//            fail(e);
//        }
    }

    @Test
    public void testHttpGetPair() {
        GetRequest getRequest = new GetRequest(String.format(API_PAIR_URL_PROPERTY, API_KEY_PROPERTY, BASE_CURRENCY, TARGET_CURRENCY));
        assertNotNull(getRequest);
        assertEquals(String.format(API_PAIR_URL, API_KEY_PROPERTY, BASE_CURRENCY, TARGET_CURRENCY), getRequest.getUrl());
        String response = getRequest.fetchResponse();
        assertNotNull(response);
        assertNotEquals(ERROR_INVALID_KEY, response);
        assertNotEquals(ERROR_NO_KEY, response);
//        try {
//            FileHelper.writeStringToFile(response, PAIR_RESPONSE_FILE);
//        } catch (IOException e) {
//            fail(e);
//        }
    }

    @Test
    public void parsePairResponse() {
        ExchangeRateApiPairResponse response = null;
        try {
            response = JsonHelper.parseResponse(new File(PAIR_RESPONSE_FILE), ExchangeRateApiPairResponse.class);
        } catch (IOException e) {
            fail(e);
        }
        assertNotNull(response);
        assertEquals("USD", response.getBase_code());
        assertEquals("ZAR", response.getTarget_code());
        assertTrue(response.getConversion_rate() > 0);
    }

    @Test
    public void parseStandardResponse() {
        ExchangeRateApiStandardResponse response = null;
        try {
            response = JsonHelper.parseResponse(new File(STANDARD_RESPONSE_FILE), ExchangeRateApiStandardResponse.class);
        } catch (IOException e) {
           fail(e);
        }
        assertNotNull(response);
        assertEquals("USD", response.getBase_code());
        Date unixDate = DateHelper.fromUnixTimestamp(response.getTime_last_update_unix());
        Date date = null;
        try {
            date = DateHelper.fromStringUtc(response.getTime_last_update_utc(), StandardDateFormat.RFC1123.getDateFormat());
        } catch (ParseException e) {
            fail(String.format("Failed to parse date: %s", response.getTime_last_update_utc()));
        }
        assertEquals(date, unixDate);
        assertNotNull(response.getConversion_rates());
        assertFalse(response.getConversion_rates().isEmpty());
    }

    @Test
    public void testExchangeApiFetcher() {
        ExchangeRateApiFetcher fetcher = new ExchangeRateApiFetcher(APP_CONFIG, BASE_CURRENCY).fetch();
        assertNotNull(fetcher);
        assertEquals(BASE_CURRENCY, fetcher.getBaseCurrency());
        assertNotNull(fetcher.getTimeLastUpdateUtc());
        assertTrue(fetcher.isRateExists(TARGET_CURRENCY));
        assertTrue(fetcher.getConversionRate(TARGET_CURRENCY) > 0);
    }



}

File: cs/exchangeapi/Money.Tests/ExchangeRateApiTest.cs
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Text.Json;
using NUnit.Framework;
using TheSoftwareGorilla.TDD.Money;
using TheSoftwareGorilla.TDD.Money.Http;
using TheSoftwareGorilla.TDD.Money.JsonHolder;

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
//TODO: Refactor tests to segregate each class into its own test class
//TODO: Add a CurrencyExchange object to cache all exchange rates and lazy load them as needed.
//TODO: Get to 100% code coverage.
//TODO: Create the utility class for the JSON parsing.
//TODO: Create a utility class for the HTTP requests.
//TODO: Build solution for error responses.
#endregion

public class ExchangeRateApiTest
{
    public const string BASE_CURRENCY = "USD";
    public const string TARGET_CURRENCY = "EUR";


    private static readonly string API_KEY_SETTING;
    private static readonly string API_PAIR_URL_SETTING;
    private static readonly string API_STANDARD_URL_SETTING;
    private static readonly AppConfig APP_SETTINGS;
    private static readonly string PAIR_RESPONSE_FILE;
    private static readonly string STANDARD_RESPONSE_FILE;
    private static readonly DateTime TEST_DATE_STANDARD;

    static ExchangeRateApiTest() {
        APP_SETTINGS = new AppConfig();
        API_KEY_SETTING = APP_SETTINGS.ApiKey;
        API_PAIR_URL_SETTING = APP_SETTINGS.ExchangeApiPairUrl;
        API_STANDARD_URL_SETTING = APP_SETTINGS.ExchangeApiStandardUrl;
        PAIR_RESPONSE_FILE = AppConfig.GetSolutionPathForFile("pair_response.json");
        STANDARD_RESPONSE_FILE = AppConfig.GetSolutionPathForFile("standard_response.json");
        TEST_DATE_STANDARD = DateTime.ParseExact("Wed, 30 Oct 2024 00:00:01 +0000", "ddd, dd MMM yyyy HH:mm:ss K", CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal);
    }

    [SetUp]
    public void Setup()
    {
    }

    [Test]
    public void TestHttpGetStandard() {
        GetRequest request = new GetRequest(String.Format(API_STANDARD_URL_SETTING, API_KEY_SETTING, BASE_CURRENCY));   
        string response = request.FetchResponseAsync().Result;
        Assert.IsNotNull(response);
    }

    [Test]
    public void TestHttpGetPair() {
        GetRequest request = new GetRequest(String.Format(API_PAIR_URL_SETTING, API_KEY_SETTING, BASE_CURRENCY, TARGET_CURRENCY));   
        string response = request.FetchResponseAsync().Result;
        Assert.IsNotNull(response);
    }

    [Test]
    public void TestDeserializePairResponse() {
        string jsonString = File.ReadAllText(PAIR_RESPONSE_FILE);
        ExchangeRateApiPairResponse response = JsonSerializer.Deserialize<ExchangeRateApiPairResponse>(jsonString);
        Assert.IsNotNull(response);
        Assert.That(response.BaseCode, Is.EqualTo(BASE_CURRENCY));
        Assert.That(response.TargetCode, Is.EqualTo(TARGET_CURRENCY));
        Assert.IsNotNull(response.ConversionRate);
        Assert.That(response.ConversionRate, Is.GreaterThan(0));
    }

    [Test]
    public void TestDeserializeStandardResponse() {
        string jsonString = File.ReadAllText(STANDARD_RESPONSE_FILE);
        ExchangeRateApiStandardResponse response = JsonSerializer.Deserialize<ExchangeRateApiStandardResponse>(jsonString);
        Assert.IsNotNull(response);
        Assert.That(response.BaseCode, Is.EqualTo(BASE_CURRENCY));
        Assert.IsNotNull(response.TimeLastUpdateUtc);
        DateTime utcDate = GetDateFromStringUtc(response.TimeLastUpdateUtc).Value;
        Assert.That(utcDate, Is.EqualTo(TEST_DATE_STANDARD));
        Assert.IsNotNull(response.TimeLastUpdateUnix);
        DateTime unixDate = GetDateFromUnixTimestamp(response.TimeLastUpdateUnix.Value);
        Assert.That(unixDate, Is.EqualTo(TEST_DATE_STANDARD));
        Assert.That(utcDate, Is.EqualTo(unixDate));
        Assert.IsNotNull(response.ConversionRates);
        Assert.That(response.ConversionRates.Count, Is.GreaterThan(0));
        Assert.That(response.ConversionRates.ContainsKey(TARGET_CURRENCY));
        Assert.That(response.ConversionRates[TARGET_CURRENCY], Is.GreaterThan(0));
    }

    [Test]
    public void TestExchangeRateApiFetcher() {
        ExchangeRateApiFetcher fetcher = new ExchangeRateApiFetcher(APP_SETTINGS, BASE_CURRENCY).Fetch();
        Assert.IsNotNull(fetcher);
        Assert.That(fetcher.BaseCurrency, Is.EqualTo(BASE_CURRENCY));
        Assert.IsNotNull(fetcher.TimeLastUpdateUtc);
        Assert.IsNotNull(fetcher.ConversionRates);
        Assert.IsNotNull(fetcher.ConversionRates[TARGET_CURRENCY]);
        Assert.That(fetcher.ConversionRates[TARGET_CURRENCY], Is.GreaterThan(0));
    } 

    private static DateTime GetLocalDateFromUtcDate(DateTime dateTime) 
    {
        DateTime localDate = TimeZoneInfo.ConvertTimeFromUtc(dateTime, TimeZoneInfo.Local);
        return localDate;
    }

    private static DateTime GetUtcDateFromLocalDate(DateTime dateTime) 
    {
        DateTime utcDate = TimeZoneInfo.ConvertTimeToUtc(dateTime, TimeZoneInfo.Utc);
        return utcDate;
    }

    private static DateTime GetLocalDateFromUnixTimestampUtc(int timestamp)
    {
        DateTimeOffset dateTimeOffset = DateTimeOffset.FromUnixTimeSeconds(timestamp);
        DateTime localDateTime = dateTimeOffset.LocalDateTime;
        return localDateTime;
    }
    private static DateTime GetDateFromUnixTimestamp(int timestamp)
    {
        DateTime response = DateTime.UnixEpoch.AddSeconds(timestamp);
        return response;
    }

    private static DateTime? GetDateFromStringUtc(string dateString) {
        const string format = "ddd, dd MMM yyyy HH:mm:ss K";
        try
        {
            DateTime date = DateTime.ParseExact(dateString, format, CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal);
            return date;
        }
        catch (FormatException e)
        {
            Console.WriteLine(e.Message);
            return null;
        }

    }       

    private static DateTime? GetLocalDateFromStringUtc(string dateString) {
        const string format = "ddd, dd MMM yyyy HH:mm:ss K";
        var localTimeZone = TimeZoneInfo.Local;

        try
        {
            DateTime date = DateTime.ParseExact(dateString, format, CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal);
            DateTime localDate = TimeZoneInfo.ConvertTimeFromUtc(date, localTimeZone);
            return localDate;
        }
        catch (FormatException e)
        {
            Console.WriteLine(e.Message);
            return null;
        }

    }       
}